### ISSUE TITLE
[Breadcrumb] Firefox ESR does not support ResizeObserver: Failed to return correct breadcrumb list and title when the breadcrumb is dynamically updated 

### ISSUE THREAD

**Issue by RogerRosado** [IBM/carbon-addons-iot-react#1158]
### Describe the bug

Firefox ESR does not support ResizeObserver: Failed to return correct breadcrumb list and title when the breadcrumb is dynamically updated (Breadcrumb has the “hasOverflow === true” which is defaulted for both Graphite and Health).

The following three (hasOverflow and browser support for ResizeObserver) combinations seems to work fine: 

1. hasOverflow === true”  and “ResizeObserver !== undefined” ( for browsers that DO support ResizeObserver)

2. hasOverflow === false” and “ResizeObserver === undefined” ( for browsers that do NOT support ResizeObserver)

3. hasOverflow is set to false and and “ResizeObserver !== undefined” ( for browsers that DO support ResizeObserver) 
          **_Note_**… for this combination it throws an incorrect Warning: 
“Warning: You have set hasOverflow to true, but the current browser does not support ResizeObserver. You will need to include a ResizeObserver polyfill for hasOverflow to function properly”

### To Reproduce

This can be validated on the Graphite Reference Application under tables:

Using Firefox ESR browser (which does not support ResizeObserver)
Click on WONUM000 and then click on “Next tab”, or back to table and click another work order. 
The breadcrumb display is incorrect:

![image](https://user-images.githubusercontent.com/56408920/81285749-fc5b6e80-902d-11ea-94a5-d25a756f8ff1.png)

### Expected behavior

![image](https://user-images.githubusercontent.com/56408920/81285786-08dfc700-902e-11ea-80c4-cc29ed32f922.png)

### Environment/versions:

- Browser: FireFox ESR or any non-support ResizeObserver browser
- `carbon-addons-iot-react` latest version: 

### Additional context

I debugged and reviewed the Breadcrumb.jsx in Iot; I believe the simplest fix is to modify line 117 from:
      {hasOverflow ? (

To:
      {hasOverflow && browserSupportsResizeObserver ? ( 

Or maybe:
      {hasOverflow && breadcrumbRef.current ? ( 

Breadcrumb.jsx code snippet:

![image](https://user-images.githubusercontent.com/56408920/81285425-7ccd9f80-902d-11ea-957d-1206ceffd7c6.png)

### Specific timeline issues / requests

Health June/July release

**Comment by davidicus** [comment]
@RogerRosado are you using the resizeObserver polyfill? The resize feature, which is turned on by the presence of the `hasOverflow` prop must include the polyfill to work on browsers that do not suppport resizeObserver natively. If you open [this](https://carbon-addons-iot-react.com/?path=/story/watson-iot-breadcrumb--hasoverflow) link you can see our resize story is working in FF.

**Comment by stuckless** [comment]
@davidicus - I do agree that FF does support the resize observer from FF 69+, but, this code is detecting that it is not there.  I think that fundamentally, the issue here is that if I set `hasOverflow=true` but there is no resize observer support then hasOverflow should be reset back to false.  ie, we should not be trying to do the resize observer stuff if we know it's not supported.  So the expectation woudl be that when resize observer is not detected that hasOverflow is simply assumed to be false, and in that way, we get a consistent behaviour.  Right now we are dropping items if we tell it hasOverflow but the overlow can't be done.

hasOverflow almost needs a re-assignment based on the fact that support for resizing must exist.
`hasOverflow  = browserSupportsResizeObserver && hasOverflow`

**Comment by davidicus** [comment]
@stuckless @RogerRosado are you not seeing this warning in your console? https://github.com/IBM/carbon-addons-iot-react/blob/3b67aa78a32d3329f1e6ba378ba17e9af020f59b/src/components/Breadcrumb/Breadcrumb.jsx#L63

**Comment by davidicus** [comment]
Even still we should check if its supported and flip the `hasOverflow` bit if its not to keep the component working as intended.

**Comment by stuckless** [comment]
> @stuckless @RogerRosado are you not seeing this warning in your console?
> 
> https://github.com/IBM/carbon-addons-iot-react/blob/3b67aa78a32d3329f1e6ba378ba17e9af020f59b/src/components/Breadcrumb/Breadcrumb.jsx#L63

Yeah, it is showing the warning, which is odd, given that it's a FF browser.  We may not ship with a resizeobserver polyfill, so the expectation would be that hasOverflow has no effect if we resizeobserver is not available.  But right now, hasOverflow with without resize observer (does give a warning) but, it also is dropping items :(

**Comment by tay1orjones** [comment]
> ... FF does support the resize observer from FF 69+, but, this code is detecting that it is not there

The browser detection function properly works in firefox. 
https://github.com/IBM/carbon-addons-iot-react/blob/b1020f2ffde5e64d5f91c1cefb459bb6a6136fd2/src/utils/componentUtilityFunctions.js#L315-L316
I took a second look at [the docs for warning](https://www.npmjs.com/package/warning#usage) - I think the warning message condition needs to be inversed, so it pops only when RO _isn't_ available. Looks like it should be fixed by:
```diff
if (__DEV__) {
    warning(
-      browserSupportsResizeObserver && hasOverflow,
-      !browserSupportsResizeObserver && hasOverflow,
      'You have set hasOverflow to true, but the current browser does not support ResizeObserver. You will need to include a ResizeObserver polyfill for hasOverflow to function properly.'
    );
  }
```

The other issue of items being dropped - some context: #1020 didn't modify the `hasOverflow` conditional render of the breadcrumb because it doesn't cause the component to unmount. It's unpredictable behavior, but the docs state that `hasOverflow` implies the availability of ResizeObserver. 

To fix - props can't be overridden, but the condition can be further qualified with what you suggest:
```
{hasOverflow && browserSupportsResizeObserver ? (
```

**Comment by RogerRosado** [comment]
I just updated FF... the MacOS FF is at 68.8.0 ESR version which comes defaulted with maxOS High Sierra:
![image](https://user-images.githubusercontent.com/56408920/81327757-c340ef80-9069-11ea-8c8f-3fe70ddc82b0.png)

It seems that I may be able to get desktop version which is at  76.0... I'll do that after this issue is resolved

**Comment by tay1orjones** [comment]
Ah, that's my point of confusion. I was looking at non ESR Firefox. ESR isn't a default for MacOS. It's a default provided for Macs administered through MAC@IBM. 

I don't believe ESR has support for ResizeObserver, so you'll need to supply a polyfill for that environment.

We shouldn't need to inverse the condition on the `warning()`, it's working as intended.

**Comment by RogerRosado** [comment]
@tay1orjones the warning is shown when you have hasOverflow = false but the browser support support ResizeObserver...

Well I guess if the browser supports ResizeObserver it why not have hasOverflow = true... but warning then should say browser supports ResizeObserver but the component hasOverflow is set to false

**Comment by tay1orjones** [comment]
Oh! You're right - this is confusing.

I think to fix this the condition should be `browserSupportsResizeObserver && hasOverflow || !hasOverflow`?

**Comment by stuckless** [comment]
> Oh! You're right - this is confusing.
> 
> I think to fix this the condition should be
> 
> ```
> browserSupportsResizeObserver && hasOverflow || !hasOverflow
> ```
> 
> ?

Are you sure?
If I have resize support AND i have overflow enabled then do `this`
or
if have resize support AND I don't have hasOverflow enabled, then do `this`

It does `this` no matter the value of hasOverflow and only changes the behavior when browser supports the resize observer, which I don't think is correct logic.

This is why I'm advocate for clear intention variables over complex conditionals, and why I feel that anytime someone uses and/or conditions together there should be a comment, since I know myself, I routinely "think" what I want but my and/or conditions don't always reflect (usually I'm missing brakcets) :)

**Comment by tay1orjones** [comment]
Closed via #1163
